### DISCUSSION TITLE
Is it possible to set up grouping or aggregation for http_client metrics ?

### DISCUSSION THREAD

**Discussion by darkangelmsk** [open-telemetry/opentelemetry-dotnet#5446] [Q&A]
Greetings! 
We noticed, what Amazon S3 ocassionaly change peeer server (and IP address) so out metrics grow huge with information we don't really need.  I tries to remove **network_peer_address** but nethertheless we still get dublicate metris, even without IPs.  

![image](https://github.com/open-telemetry/opentelemetry-dotnet/assets/98049821/eb2fce56-5d00-4465-8584-3fe0974888e4)

Is it possible to somehow setup grouping for metrict not to take IP address label into account ?  Or somehow to filter metrics for urls contaings "amazonaws' ?

P.S. We use .Net 8 version of library

Thank you!

**Comment by vishweshbankwar** [comment]
@darkangelmsk - You can use [views](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/metrics/customizing-the-sdk#select-specific-tags) for selecting particular dimensions within a metric. Looks like you are already doing that for `network_peer_address`?

> somehow to filter metrics for urls contaings "amazonaws'

This isn't possible today.

**Reply by darkangelmsk** [reply]
@vishweshbankwar 

First I tried to absolutely remove all records about Amazon AWS, using thsi code. But this changed nothing, I still get a bunch of record for .amazonaws.com with different IPs:
`
            .WithTracing(tracing => tracing
            .AddHttpClientInstrumentation(
                // Note: Only called on .NET & .NET Core runtimes.
                (options) => options.FilterHttpRequestMessage =
                    (httpRequestMessage) =>
                    {
                        return !(httpRequestMessage.RequestUri != null && httpRequestMessage.RequestUri.AbsoluteUri.Contains("amazonaws");
                    })
` 

When I switched to views, to get rid of IPs, but I still get several records for amazon (and ya.ru - I manually spoofed an IP for it via hosts file) with different IPs, with the only difference - there's no IP adresses in metrics output. 
Code I used is following

`
            .AddMeter("System.Net.Http")
            .AddView(
                instrumentName: "http.client.connection.duration",
                metricStreamConfiguration: new MetricStreamConfiguration
                {
                    TagKeys = new string[] { "url.scheme", "network.protocol.version", "server.address", "server.port" }
                }
             )
            .AddView(
                instrumentName: "http.client.open_connections",
                metricStreamConfiguration: new MetricStreamConfiguration
                {
                    TagKeys = new string[] { "http.connection.state", "url.scheme", "network.protocol.version", "server.address", "server.port" }
                }
             )
`

So, I faced two questions:
1) Is there any solution to output only one metric for all ".amazonaws.com", no matter that IP is?

2) Is it possible to filter and don't record any http activity for any URL containing "amazonaws" ?

That you for your answer/

**Reply by vishweshbankwar** [reply]
@darkangelmsk 

> Is there any solution to output only one metric for all ".amazonaws.com", no matter that IP is?

Sorry, I am not following. Do you mind elaborating this bit more? 

Filtering/aggregating metrics based on dimension values is not possible, if that's what you are trying.

> Is it possible to filter and don't record any http activity for any URL containing "amazonaws"?

I see you are already using Filter API for this. This should work, if it is not then please share a minimal repro app so that we can take a look.

**Reply by acu-kevinb** [reply]
@vishweshbankwar 
The issue is that if the downstream service has high cardinality for any of the tags on http.client.open_connections, it results in a large number of 'empty' telemetry items being emitted for that counter. 

In my case, just calling microsoft graph results in over 200 http.client.open_connections metrics emitted per minute when using Azure Monitor OpenTelemetry, which polls once per minute. 156 of those are just for login.microsoftonline, which has both a large number of IP addresses, and seems to switch arbitrarily between HTTP 2 and 1.1 depending on endpoint. That's 200 metrics items per minute, every minute, for the entire lifetime of my app, regardless of whether there actually is an open connection to the given endpoint or not, because the counter for a given set of tags is never removed, just set to 0.

Since there appears to be no way to filter these metrics out based on tag value, I had to resort to disabling all the metrics from System.Net.Http, which is far from ideal.
